src/pushprocessor.ts

```diff
@@ -162,24 +168,31 @@ const EXPECTED_DEFAULT_UNDERRIDE_RULE_IDS = [
 /**
  * Make sure that each of the rules listed in `defaultRuleIds` is listed in the given set of push rules.
  *
+ * @param kind - the kind of push rule set being merged.
  * @param incomingRules - the existing set of known push rules for the user.
  * @param defaultRules - a lookup table for the default definitions of push rules.
  * @param defaultRuleIds - the IDs of the expected default push rules, in order.
  *
  * @returns A copy of `incomingRules`, with any missing default rules inserted in the right place.
  */
 function mergeRulesWithDefaults(
+    kind: PushRuleKind,
     incomingRules: IPushRule[],
     defaultRules: Record<string, IPushRule>,
-    defaultRuleIds: string[],
+    defaultRuleIds: OrderedRules,
 ): IPushRule[] {
-    // Calculate the index after the last default rule in `incomingRules`
-    // to allow us to split the incomingRules into defaults and custom
-    let firstCustomRuleIndex = incomingRules.findIndex((r) => !r.default);
-    if (firstCustomRuleIndex < 0) firstCustomRuleIndex = incomingRules.length;
-
-    function insertDefaultPushRule(ruleId: string): void {
-        if (ruleId in defaultRules) {
+    // Find the indices of the edges of the user-defined rules in the incoming rules
+    const incomingRulesEnabled = incomingRules.map((rule) => rule.enabled);
```

Comment:
why are we looking at `enabled` to find user-defined rules, rather than `default`?

It's possible for default rules to be disabled, so this seems unreliable